---
title: "RNA-seq data normalization"
subtitle: "Layla Saad Montoya internship"
author: " Florent Chuffart"
date: "`r Sys.Date()`"
output: 
  rmarkdown::html_document:
    toc: true
    toc_float: true
    toc_depth: 3
    number_sections: true
---

# Introduction

## Libraries

```{r, echo=FALSE, eval=TRUE, label="Loading libraries . . ."}
setwd("C:/Users/lasam/Desktop/stage_iab/sujets/nttcga_study")
knitr::opts_chunk$set(collapse=TRUE, comment = "#>", fig.width=9, fig.height=6, eval=TRUE, echo=FALSE, results="hide", dpi=75)
```

## Data

### Features

File containing the information on every gene in the geneset.

```{r, label = "Loading file . . .", results='asis'}
features_filename = "features_nttcga.rds"
features = readRDS(features_filename)
```

```{r}
complex = c(#Lung
            "ROR1-AS1" = 1,
            "CHIAP2" = 1,
            "CHIA" = 1,
            "LOC101927884" = 1,
            "SNTN" = 1,
            "LINC02016" = 1,
            "HTR3C" = 1,
            "CSF2" = 1,
            "ROS1" = 1,
            "RPL13AP17" = 1,
            "DCSTAMP" = 1,
            "MS4A15" = 1,
            "MIR614" = 1,
            "LMO7DN-IT1" = 1,
            "LINC00551" = 1,
            "MYO16-AS1" = 1,
            "NKX2-1-AS1" = 1,
            "FAM92B" = 1,
            "LOC101928418" = 1,
            "PARAL1" = 1,
            "CT45A2" = 1,
            "CT45A8" = 1,
            "CT45A9" = 1,
            "CT45A10" = 1,
            #kidney
            "CLDN19" = 2,      
            "LINC01788" = 2,    
            "LINC01762" = 2,    
            "ATP6V1G3" = 2,     
            "OR2T10" = 2,       
            "LINC01158" = 2,   
            "SLC22A13" = 2,     
            "TMEM207" = 2,      
            "LOC100506444" = 2, 
            "AFP" = 2,          
            "SLC6A18" = 2,      
            "LINC02149" = 2,   
            "CDH9" = 2,         
            "TMEM174" = 2,      
            "SLC4A9" = 2,       
            "SLC34A1" = 2,      
            "PFN3" = 2,         
            "MCCD1" = 2,       
            "SLC22A2" = 2,      
            "LINC01510" = 2,    
            "SLC7A13" = 2,      
            "LINC01230" = 2,    
            "LINC00845" = 2,    
            "HMX2" = 2,        
            "SLC22A24" = 2,     
            "SLC22A11" = 2,     
            "SLC22A12" = 2,     
            "KRTAP5-8" = 2,     
            "KCNJ1" = 2,        
            "ENOX1-AS2" = 2,  
            "LINC00379" = 2,    
            "LINC02294" = 2,    
            "LINC00645" = 2,    
            "YWHAEP7" = 2,      
            "RDH8" = 2,         
            "APCDD1L-AS1" = 2, 
            "SMIM34B" = 2,      
            "DNMT3L" = 2,       
            "COL18A1-AS1" = 2,  
            "PLAC1" = 2,
            #liver
            "PRAMEF10" = 3,
            "PRAMEF33" = 3,
            "C8A" = 3,
            "NBPF13P" = 3,
            "CFHR4" = 3,
            "CFHR2" = 3,
            "CFHR5" = 3,
            "F13B" = 3,
            "BMP10" = 3,
            "CPS1-IT1" = 3,
            "SPP2" = 3,
            "UROC1" = 3,       
            "LOC100507389" = 3,
            "LINC02037" = 3, 
            "LINC02428" = 3, 
            "LINC02159" = 3, 
            "SLC17A2" = 3,    
            "SLC22A1" = 3,     
            "POU6F2-AS1" = 3,  
            "LOC157273" = 3,   
            "CYP7A1" = 3,      
            "AKR1C6P" = 3,    
            "GDF2" = 3,        
            "MBL2" = 3,        
            "F2" = 3,          
            "SLC22A25" = 3,     
            "SLC22A10" = 3,    
            "SLC22A9" = 3,     
            "LOC101929427" = 3,
            "SLCO1B1" = 3,     
            "INHBE" = 3,       
            "SERPINA10" = 3,   
            "SERPINA2" = 3,    
            "CYP1A2" = 3,      
            "NPW" = 3,         
            "MT1B" = 3,        
            "CA5A" = 3,         
            "LOC101927571" = 3, 
            "C3P1" = 3,         
            "CYP2A7" = 3,       
            "CYP2A13" = 3,      
            "APOC4" = 3,       
            "FGF21" = 3,       
            "HAO1" = 3,         
            "LINC01370" = 3,    
            "SERPINA7" = 3,     
            "F9" = 3,           
            "TTTY2B" = 3,      
            "TTTY2" = 3,
            #prostate
            "CYP4F62P" = 4,
            "CYP4F30P" = 4,     
            "LOC440910" = 4,    
            "PCGEM1" = 4,       
            "SOX14" = 4,        
            "PCAT4" = 4,        
            "SP8" = 4,          
            "RLN1" = 4,         
            "PCA3" = 4,        
            "MKX-AS1" = 4,      
            "LOC283194" = 4,    
            "LINC02297" = 4,    
            "LOC100508046" = 4, 
            "LOC101929572" = 4, 
            "OACYLP" = 4,       
            "CYP4F8" = 4,       
            "SEMG1" = 4,        
            "LINC00314" = 4,   
            "LINC00161" = 4,    
            "DSCAM-AS1" = 4,    
            "POTEH-AS1" = 4)
```

### TCGA studies

```{r study, echo = TRUE, results="verbatim"}
study_nttcga_filename = "study_nttcga_trscr.rds"
if (!exists("study_nttcga")) {
  print("Loading study_nttcga...")
  study_nttcga = readRDS(study_nttcga_filename)  
}
s = study_nttcga
idx_features = intersect(rownames(features), rownames(s$data))
print(table(study_nttcga$exp_grp$main_gse_number, study_nttcga$exp_grp$tissue_status))
```

### Gene expression profiles


```{r, fig.height=3}
plot.new()
legend("bottomleft", col=c("orange", "yellow", "green", "red"), pch=15, c("complex 1", "complex 1 & 2", "complex 2"), bty="n")
legend("bottomright", col=c("grey20", "grey80"), pch=15, c("normal", "tumoral"), bty="n")
```

### Raw counts dataset

```{r echo=TRUE, collapse=TRUE, results='asis'}
raw_counts <- s$stuffs$raw_counts
dim(raw_counts)
knitr::kable(raw_counts[1:6,1:4], align = "c", caption = "Raw counts: 6 genes × 4 samples")
```

# Data Normalisation 

## RPM

RPM (or CPM) is a common normalization method based on a transcript expression unit. It accounts for biases due to sequencing depth and is commonly used for cutoff in independent filtering.

### Formula / script

To calculate RPM the formula consists of dividing the number of raW counts by the total number of reads (or library size) and multipliying by a 1 million factor.

EdgeR has it's own cpm-calculating function included in the package. This function calulates the log-cpm of counts. However, these methods don't differ much.

In this study we calculate RPM using EdgeR's cpm() function. This method differs from standard RPM calculation because it calculates the log-cpm.

Log RPM: 

```{r, eval=TRUE, echo=TRUE}
cpm <- edgeR::cpm(raw_counts)
cpm[1:6, 1:6]
```

Standard RPM: 

```{r, eval=TRUE, echo=TRUE}
rpm_sf <- colSums(raw_counts)/10^6
rpm <- t(t(raw_counts)/rpm_sf)
(rpm)[1:6, 1:6]
```

Here we observe that both log rpm and standard rpm counts are very similar.

```{r, eval=TRUE, echo=TRUE}
plot(log2(rpm[,1]+1), log2(cpm[,1]+1))
```

### Normalized Data

```{r echo=FALSE, collapse=TRUE, results='asis'}
knitr::kable(rpm[1:6,1:4], align = "c", caption = "RPM: 6 genes × 4 samples")
```

### Heatmaps

```{r, eval=TRUE, echo=FALSE, fig.height = 9}
data <- log2(rpm+1)[idx_features,]
rpm_hm = dmethr::plot_expr_hm(data, 
  normalization="zscore_rows", 
  hcmeth_rows="cor"                ,
  csc=c("grey20", "grey80")[as.factor(s$exp_grp[colnames(data),]$tissue_status)],
  rsc=c("orange", "green", "yellow")[complex[rownames(data)]],
  main = "RPM")
```

### Clusters

RPM Clusters: 

```{r, eval=TRUE, echo=TRUE}
bar <- cutree(rpm_hm$hc_row, k = 4)
bar

names(bar)

rpm_clust_A = names(bar)[bar==1]
rpm_clust_B = names(bar)[bar==2]
rpm_clust_C = names(bar)[bar==3]
rpm_clust_D = names(bar)[bar==4]

```

<!-- ## RPKM -->

<!-- ### Formula/Script -->


<!-- ```{r} -->
<!-- rpkm = s$stuffs$rpkm_data -->
<!-- ``` -->


<!-- ### Normalized Data -->

<!-- ```{r echo=FALSE, collapse=TRUE, results='asis'} -->
<!-- knitr::kable(rpkm[1:6,1:4], align = "c", caption = "RPKM: 6 genes × 4 samples") -->
<!-- ``` -->

<!-- ### Heatmap -->

<!-- ```{r, fig.height=9} -->
<!-- data <- log2(rpkm+1)[idx_features,] -->
<!-- rpkm_hm = dmethr::plot_expr_hm(data,  -->
<!--   normalization="zscore_rows",  -->
<!--   hcmeth_rows="cor"                , -->
<!--   csc=c("grey20", "grey80")[as.factor(s$exp_grp[colnames(data),]$tissue_status)], -->
<!--   rsc=c("orange", "green", "yellow")[complex[rownames(data)]], main = "RPKM") -->
<!-- ``` -->

<!-- ### Clusters -->

<!-- Because RPKM takes gene langth into account, we won't be focusing on clustering using this method. --- Can't calculate size factor --- -->

## DESeq2

DESeq2 is a standard normalization method used for differential expression analysis. It is based on the RLE method (Relative Log Expression) and accounts for biases such as sequencing depth, library composition and PCR affinity / RNA composition.

### Formula/Script

```{r, eval=TRUE, echo=TRUE}
countData <- raw_counts
colData <- data.frame(id=colnames(countData))
dds <- DESeq2::DESeqDataSetFromMatrix(countData=countData, colData=colData, design= ~ id)
dds <- DESeq2::estimateSizeFactors(dds)
deseq2_sf <- dds$sizeFactor
deseq2 <- DESeq2::counts(dds, normalized=TRUE)
```

### Normalized Data

```{r echo=FALSE, collapse=TRUE, results='asis'}
knitr::kable(deseq2[1:6,1:4], align = "c", caption = "DESeq2: 6 genes × 4 samples")
```

### Heatmap

```{r, fig.height=9}
data <- log2(deseq2+1)[idx_features,]
deseq2_hm = dmethr::plot_expr_hm(data, 
  normalization="zscore_rows", 
  hcmeth_rows="cor"                ,
  csc=c("grey20", "grey80")[as.factor(s$exp_grp[colnames(data),]$tissue_status)],
  rsc=c("orange", "green", "yellow")[complex[rownames(data)]], main = "DESeq2")
```

### Clusters

DESeq2 Clusters: 

```{r, eval=TRUE, echo=TRUE}
bar <- cutree(deseq2_hm$hc_row, k = 4)
bar

names(bar)

deseq2_clust_A = names(bar)[bar==1]
deseq2_clust_B = names(bar)[bar==2]
deseq2_clust_C = names(bar)[bar==3]
deseq2_clust_D = names(bar)[bar==4]
```

## EdgeR

EdgeR is a more robust method that competes with DESeq2. EdgeR uses a TMM method to normalize counts (Trimmed Mean of M-values) and accounts for the same biases as DESeq2 such as library composition, sequencing depth and PCR affinity / RNA composition.TMM is a method that calculates size factors by first selecting a reference sample then calculating pseudo size factors using the log of the reference_sample over current sample ratio, and finally trimming the most extreme ratios and most influential genes. This way, EdgeR only takes into account differentially expressed genes and ignores the influence of tissue-specific genes.

### Formula/Script

```{r, eval=TRUE, echo=TRUE}
edgeRnorm <- function(rc){
    nf <- edgeR::calcNormFactors(rc)
    nc <- rc
    for(i in 1:length(nf)){
        nc[,i] <- nc[,i]*nf[i]
    }
    return(nc)
}
data <- raw_counts
edger <- edgeRnorm(data)
s$stuffs$edger_data <- edger
edger_sf <- edgeR::calcNormFactors(data, method = "TMM")
```

### Normalized Data

```{r echo=FALSE, collapse=TRUE, results='asis'}
knitr::kable(edger[1:6,1:4], align = "c", caption = "EdgeR: 6 genes × 4 samples")
```

### Heatmap

```{r, eval=TRUE, echo=TRUE, fig.height = 9}
data <- log2(edger)[idx_features,]
edger_hm = dmethr::plot_expr_hm(data, 
  normalization="zscore_rows", 
  hcmeth_rows="cor"                ,
  csc=c("grey20", "grey80")[as.factor(s$exp_grp[colnames(data),]$tissue_status)],
  rsc=c("orange", "green", "yellow")[complex[rownames(data)]],
  main = "EdgeR")
```

EdgeR Clusters

```{r, eval=TRUE, echo=TRUE}
bar <- cutree(edger_hm$hc_row, k = 4)
bar

names(bar)

edger_clust_A = names(bar)[bar==1]
edger_clust_B = names(bar)[bar==2]
edger_clust_C = names(bar)[bar==3]
edger_clust_D = names(bar)[bar==4]
```

## RPM40

### Formula

### Data

```{r, eval=TRUE, echo=TRUE}
rpm40_sf <- (colSums(raw_counts)/10^6)*40
rpm40 <- t(t(raw_counts)/rpm40_sf)
(rpm40)[1:6, 1:6]
#(rpm)[1:6, 1:6]
```

### Heatmap

```{r, eval=TRUE, echo=FALSE, fig.height = 9}
data <- log2(rpm40+1)[idx_features,]
rpm40_hm = dmethr::plot_expr_hm(data, 
  normalization="zscore_rows", 
  hcmeth_rows="cor"                ,
  csc=c("grey20", "grey80")[as.factor(s$exp_grp[colnames(data),]$tissue_status)],
  rsc=c("orange", "green", "yellow")[complex[rownames(data)]],
  main = "RPM40")
```

### Clusters

```{r, eval=TRUE, echo=TRUE}
bar <- cutree(rpm40_hm$hc_row, k = 4)
bar

names(bar)

rpm40_clust_A = names(bar)[bar==1]
rpm40_clust_B = names(bar)[bar==2]
rpm40_clust_C = names(bar)[bar==3]
rpm40_clust_D = names(bar)[bar==4]
```

# Comparison

We create a tab in which we input the clustering data.And we create a data-frame with size factors of each method.

### Plots

```{r, fig.height=9}
df_sf <- cbind(rpm_sf = rpm_sf, deseq2_sf=deseq2_sf, edger_sf=edger_sf)
dim(df_sf)
pairs(df_sf)

epimedtools::et_pairs(df_sf)
# INSTALL!!!!

```

```{r, fig.height=9}
df_sample1 <- cbind(rpm_sample1=rpm[,1], deseq2_sample1=deseq2[,1], edger_sample1=edger[,1])
dim(df_sample1)
pairs(log2(df_sample1+1))

# epimedtools::et_pairs(df_sf)
# INSTALL!!!!

```

```{r}

tmptab <- s$platform[idx_features,]

tmptab$rpm_clust = NA
tmptab$rpm40_clust = NA
tmptab$deseq2_clust = NA
tmptab$edger_clust = NA

# RPM

tmptab[rpm_clust_A,]$rpm_clust = "rpm_A"
tmptab[rpm_clust_B,]$rpm_clust = "rpm_B"
tmptab[rpm_clust_C,]$rpm_clust = "rpm_C"
tmptab[rpm_clust_D,]$rpm_clust = "rpm_D"

# DESeq2

tmptab[deseq2_clust_A,]$deseq2_clust = "deseq2_A"
tmptab[deseq2_clust_B,]$deseq2_clust = "deseq2_B"
tmptab[deseq2_clust_C,]$deseq2_clust = "deseq2_C"
tmptab[deseq2_clust_D,]$deseq2_clust = "deseq2_D"

# EdgeR

tmptab[edger_clust_A,]$edger_clust = "edger_A"
tmptab[edger_clust_B,]$edger_clust = "edger_B"
tmptab[edger_clust_C,]$edger_clust = "edger_C"
tmptab[edger_clust_D,]$edger_clust = "edger_D"

# RPM40

tmptab[rpm40_clust_A,]$rpm40_clust = "rpm40_A"
tmptab[rpm40_clust_B,]$rpm40_clust = "rpm40_B"
tmptab[rpm40_clust_C,]$rpm40_clust = "rpm40_C"
tmptab[rpm40_clust_D,]$rpm40_clust = "rpm40_D"

```

### Cluster Tables

```{r, results='asis'}

rpm_deseq2_clust_table <- table(tmptab$rpm_clust, tmptab$deseq2_clust)
rpm_edger_clust_table <- table(tmptab$rpm_clust, tmptab$edger_clust)
edger_deseq2_clust_table <- table(tmptab$edger_clust, tmptab$deseq2_clust)
rpm40_deseq2_clust_table <- table(tmptab$rpm40_clust, tmptab$deseq2_clust)
rpm40_edger_clust_table <- table(tmptab$rpm40_clust, tmptab$edger_clust)

knitr::kable(rpm_deseq2_clust_table, align = "c", caption = "RPM vs DESeq2")
knitr::kable(rpm_edger_clust_table, align = "c", caption = "RPM vs EdgeR")
knitr::kable(edger_deseq2_clust_table, align = "c", caption = "EdgeR VS DESeq2")
knitr::kable(rpm40_deseq2_clust_table, align = "c", caption = "RPM40 vs DESeq2")
knitr::kable(rpm40_deseq2_clust_table, align = "c", caption = "RPM40 vs EdgeR")

```

```{r, eval=TRUE, echo=TRUE}
#layout(matrix(1:3,1), respect = TRUE)
plot(edger_sf,rpm_sf, main = "EdgeR vs RPM (Size Factors)", xlab = "EdgeR TMM Size Factors", ylab = "RPM Size Factors")
dim(edger)
plot(log2(edger+1)[,1], log2(rpm+1)[,1], main = "EdgeR vs RPM: inter-sample", xlab = "EdgeR TMM Counts", ylab = "RPM Counts")
```

In this case, there is also a 40-fold difference between RPM and EdgeR size factors

```{r, eval=TRUE, echo=TRUE}
plot(log2(edger+1)[,1], log2(rpm*40+1)[,1], main = "EdgeR vs RPM40: inter-sample (corrected)", xlab = "log2 EdgeR TMM Counts", ylab = "log2 RPM40 Counts")
```

### Cluster comparison table

## DESeq2 vs EdgeR

### Plot

```{r, eval=TRUE, echo=TRUE}
# layout(matrix(1:3,1), respect = TRUE)
plot(edger_sf,deseq2_sf, main = "EdgeR vs DESeq2 (Size Factors)", xlab = "EdgeR TMM Size Factors", ylab = "DESeq2 Size Factors")
dim(edger)

plot(log2(edger+1)[,1], log2(deseq2+1)[,1], main = "EdgeR vs RPM: inter-sample", xlab = "log2 EdgeR TMM Counts", ylab = "log2 RPM Counts")
```

# Conclusion and Discussion

# References

## Articles

The portrait of liver cancer is shaped by mitochondrial genetics


## Session Info

```{r, eval=TRUE, echo=TRUE}
sessionInfo()
```
