---
title: "Impact of RNA-seq count normalization on clustering of co-expressed genes"
author: "Layla Saad Montoya and Florent Chuffart"
date: "`r Sys.Date()`"
output: 
  rmarkdown::html_document:
    toc: true
    toc_float: true
    toc_depth: 3
    number_sections: true
---

```{r, echo=FALSE, eval=TRUE, label="Loading libraries . . ."}
knitr::opts_chunk$set(collapse=TRUE, comment = "#>", fig.width=9, fig.height=6, eval=TRUE, echo=FALSE, results="hide", dpi=75)
```


# Introduction

The goal of this study is to first identify then quantify the differences between normalization methods and their subsequent gene clustering. We wish to highlight the mathematical basis for these differences as well as the ways in which these methods affect the raw results.

# Data

## Dataset

The dataset `raw_counts` is obtained from the LIHC group of TCGA database. TCGA (The Cancer Genome Atlas) is a worldwide database containing sequenced cell and tissue samples, and their respective genomic, epigenomic, transcriptomic, and proteomic data. Clincal data as well as survival data and follow-up are equally available. The  data in this group is derived from mitochondrial (?) liver samples both tumorous and non-tumorous.

```{r raw_counts, echo=TRUE, results="verbatim"}
load("data/lihc_raw_counts.RData")
raw_counts = lihc_raw_counts

# study_lihc_filename = "tcga_studies/study_TCGA-LIHC_trscr.rds"
# study_lihc = readRDS(study_lihc_filename)
# lihc_raw_counts = study_lihc$stuffs$raw_counts
# raw_counts = lihc_raw_counts

head(raw_counts[,1:4])
dim(raw_counts)
```

### Geneset

Our data, containing 25360, will be reduced to a geneset of 30 genes coding for the CPSF3 complex, involved in rna cleavage and polyadenylation, two fundamental steps in the maturation process of pre-mRNA. 

```{r echo=TRUE}
features = c(
  SSU72    = 1,  #    SSU72    29101
  LSM10    = 2,  #    LSM10    84967
  SNRPE    = 2,  #    SNRPE     6635
  CPSF3    = 3,  #    CPSF3    51692
  SNRPG    = 2,  #    SNRPG     6637
  WDR33    = 1,  #    WDR33    55339
  SLBP     = 2,  #     SLBP     7884
  FIP1L1   = 1,  #   FIP1L1    81608
  LSM11    = 2,  #    LSM11   134353
  CASP8AP2 = 2,  # CASP8AP2     9994
  CPSF4    = 1,  #    CPSF4    10898
  ERI1     = 2,  #     ERI1    90459
  CPSF1    = 1,  #    CPSF1    29894
  CSTF3    = 1,  #    CSTF3     1479
  CLP1     = 1,  #     CLP1    10978
  CPSF7    = 1,  #    CPSF7    79869
  PCF11    = 1,  #    PCF11    51585
  CPSF6    = 1,  #    CPSF6    11052
  SNRPF    = 2,  #    SNRPF     6636
  PABPN1   = 1,  #   PABPN1     8106
  CPSF2    = 3,  #    CPSF2    53981
  PAPOLA   = 1,  #   PAPOLA    10914
  RBBP6    = 1,  #    RBBP6     5930
  NUDT21   = 1,  #   NUDT21    11051
  SYMPK    = 3,  #    SYMPK     8189
  ZNF473   = 2,  #   ZNF473    25888
  SNRPB    = 2,  #    SNRPB     6628
  CSTF1    = 1,  #    CSTF1     1477
  SNRPD3   = 2,  #   SNRPD3     6634
  CSTF2    = 3   #    CSTF2     1478
)
idx_features = names(features)
```

```{r, fig.height=3}
plot.new()
legend("bottomleft", col=c("orange", "yellow", "green"), pch=15, c("complex 1", "complex 1 & 2", "complex 2"), bty="n")
legend("bottomright", col=c("grey20", "grey80"), pch=15, c("tumoral", "normal"), bty="n")
```

# RPM

Reads Per Million (RPM) is a transcript expression unit. To calculate RPM the formula consists of dividing the number of raw counts by the total number of reads (or library size) and multipliying by a 1 million factor. This method is commonly used for inter-sample comparisons and as a basic unit of gene expression. This method does not center counts around zero.

## Scaling factors

```{r, eval=TRUE, echo=TRUE}
rpm_sf = colSums(raw_counts)/10^6
rpm = t(t(raw_counts)/rpm_sf)
```

## Heatmap

```{r, eval=TRUE, echo=FALSE, fig.height = 9}
data = log2(rpm+1)[idx_features,]
rpm_hm = dmethr::plot_expr_hm(data, 
  normalization="zscore_rows", 
  hcmeth_rows="cor"                ,
  csc=c("grey20", "grey80")[as.numeric(as.factor(substr(colnames(raw_counts), 14,100)%in%"11A"))],
  rsc=c("orange", "green", "yellow")[features[rownames(data)]],
  main = "RPM")
```

```{r}
## Clusters
gene_cluster_tab = data.frame(features)
bar = cutree(rpm_hm$hc_row, k=4)
bar
gene_cluster_tab$rpm_clust = NA
gene_cluster_tab[names(bar)[bar==1],"rpm_clust"] = "rpmA"
gene_cluster_tab[names(bar)[bar==2],"rpm_clust"] = "rpmB"
gene_cluster_tab[names(bar)[bar==3],"rpm_clust"] = "rpmC"
gene_cluster_tab[names(bar)[bar==4],"rpm_clust"] = "rpmD"
```

# DESeq2

DESeq2 is a standard normalization method used for differential expression analysis. It is based on the RLE method (Relative Log Expression) and accounts for biases such as sequencing depth, library composition and PCR affinity / RNA composition. DESeq2 is the main normalisation method used in Differential Expression Analysis at the IAB.


## Scaling factors

```{r, eval=TRUE, echo=TRUE}
colData = data.frame(id=colnames(raw_counts))
dds = DESeq2::DESeqDataSetFromMatrix(countData=raw_counts, colData=colData, design= ~ 1)
dds = DESeq2::estimateSizeFactors(dds)
deseq2_sf = dds$sizeFactor
deseq2 = t(t(raw_counts)/deseq2_sf)
deseq2 = DESeq2::counts(dds, normalized=TRUE)
```

## Heatmap

```{r, fig.height=9}
data = log2(deseq2+1)[idx_features,]
deseq2_hm = dmethr::plot_expr_hm(data, 
  normalization="zscore_rows", 
  hcmeth_rows="cor"                ,
  # hcmeth_cols = "cor",
  csc=c("grey20", "grey80")[as.numeric(as.factor(substr(colnames(raw_counts), 14,100)%in%"11A"))],
  rsc=c("orange", "green", "yellow")[features[rownames(data)]], main = "DESeq2")
```

```{r}
## Clusters
bar = cutree(deseq2_hm$hc_row, k = 4)
bar
gene_cluster_tab$deseq2_clust = NA
gene_cluster_tab[names(bar)[bar==1],"deseq2_clust"] = "deseq2A"
gene_cluster_tab[names(bar)[bar==2],"deseq2_clust"] = "deseq2B"
gene_cluster_tab[names(bar)[bar==3],"deseq2_clust"] = "deseq2C"
gene_cluster_tab[names(bar)[bar==4],"deseq2_clust"] = "deseq2D"
```

# RPM vs. DESeq2

We proceed to compare both RPM and DESeq2 methods against each other in 3 key aspects: Cluster consistency, size factor correlation and raw value transformation.

## Cluster consistency table

```{r results="verbatim"}
table(gene_cluster_tab$rpm_clust, gene_cluster_tab$deseq2_clust)
```

When compared to each other, RPM and DESeq2 clustering presents us with only two consistent cluster group and one exclusive cluster group corresponding to that of `rpmD` or `deseq2D`. Other clusters contain should contain shared or transferred genes.

## Size factor correlation

```{r}
layout(matrix(1:2,1), respect=TRUE)
plot(deseq2_sf, rpm_sf)
abline(a=0, b=40, col=2)
legend("bottomright", "y=40x", col=2, lty=1)
plot(deseq2_sf, rpm_sf/40)
abline(a=0, b=1, col=2)
legend("bottomright", "y=x", col=2, lty=1)
```

In the first plot (left), size factors present a dense and correlated group with a 40-fold factor. This difference is corrected in the second plot (right). 

## Raw value transformation (per sample)

```{r}
layout(matrix(1:2,1), respect=TRUE)
plot(log2(deseq2[,1]+1), log2(rpm[,1]+1))
plot(log2(deseq2[,1]+1), log2(40*rpm[,1]+1))
```

A noticeably "crushing" effect is shown in the first plot (left). This crushing effect is caused by the RPM normalization method and affects small raw count values.

# RPM40

We obtain RPM40 counts by dividing the standard RPM size factor by 40: (total_exon_mapped_genes/1M) / 40

```{r, eval=TRUE, echo=TRUE}
rpm40_sf = rpm_sf/40
rpm40 = t(t(raw_counts)/rpm40_sf)
rpm40[1:6, 1:4]
```

## Heatmap

```{r, eval=TRUE, echo=FALSE, fig.height = 9}
data = log2(rpm40+1)[idx_features,]
rpm40_hm = dmethr::plot_expr_hm(data, 
  normalization="zscore_rows", 
  hcmeth_rows="cor"                ,
  csc=c("grey20", "grey80")[as.numeric(as.factor(substr(colnames(raw_counts), 14,100)%in%"11A"))],
  rsc=c("orange", "green", "yellow")[features[rownames(data)]],
  main = "RPM40")
```

```{r}
## Clusters
bar = cutree(rpm40_hm$hc_row, k=4)
bar
gene_cluster_tab$rpm40_clust = NA
gene_cluster_tab[names(bar)[bar==1],"rpm40_clust"] = "rpm40A"
gene_cluster_tab[names(bar)[bar==2],"rpm40_clust"] = "rpm40B"
gene_cluster_tab[names(bar)[bar==3],"rpm40_clust"] = "rpm40C"
gene_cluster_tab[names(bar)[bar==4],"rpm40_clust"] = "rpm40D"
```

# RPM40 vs. DESeq2

We compared RPM40 and DESeq2 clusters
```{r results="verbatim"}
table(gene_cluster_tab$rpm40_clust, gene_cluster_tab$deseq2_clust)
```

After RPM40 normalization, clustering is still overall inconsistent, however, loose genes (a.k.a: genes moving in between clusters) seem to be less distributed in "outlying" groups and the `deseq2D` / `rpm40D` cluster group remains consistent.

# EdgeR

EdgeR is a more complex method that competes with DESeq2. EdgeR uses a TMM method to normalize counts (Trimmed Mean of M-values) and accounts for the same biases as DESeq2 such as library composition, sequencing depth and PCR affinity / RNA composition.TMM is a method that calculates size factors by first selecting a reference sample then calculating pseudo size factors using the log of the reference_sample over current sample ratio, and finally trimming the most extreme ratios and most influential genes. This way, EdgeR only takes into account differentially expressed genes and ignores the influence of tissue-specific genes.

## Scaling factors

```{r, eval=TRUE, echo=TRUE}
edger_sf = edgeR::calcNormFactors(raw_counts, method="TMM")
edger = t(t(raw_counts)*edger_sf)
```

## Heatmap

```{r, eval=TRUE, echo=TRUE, fig.height = 9}
data = log2(edger+1)[idx_features,]
edger_hm = dmethr::plot_expr_hm(data, 
  normalization="zscore_rows", 
  hcmeth_rows="cor"                ,
  csc=c("grey20", "grey80")[as.numeric(as.factor(substr(colnames(raw_counts), 14,100)%in%"11A"))],
  rsc=c("orange", "green", "yellow")[features[rownames(data)]],
  main = "EdgeR")
```

```{r}
## Clusters
bar = cutree(edger_hm$hc_row, k = 4)
bar
gene_cluster_tab$edger_clust = NA
gene_cluster_tab[names(bar)[bar==1],"edger_clust"] = "edgerA"
gene_cluster_tab[names(bar)[bar==2],"edger_clust"] = "edgerB"
gene_cluster_tab[names(bar)[bar==3],"edger_clust"] = "edgerC"
gene_cluster_tab[names(bar)[bar==4],"edger_clust"] = "edgerD"
```

# Comparison

We create a tab in which we input the clustering data. And we create a data-frame with size factors of each method.

## Scaling factors correlation plot

```{r, fig.height=9}
df_sf = cbind(rpm_sf=rpm_sf, deseq2_sf=deseq2_sf, rpm40_sf=rpm40_sf, edger_sf=edger_sf)
epimedtools::et_pairs(df_sf)
```

Scaling factors computed using RPM, RPM40 and DESeq2 appear to have a high correlation (R²=0.67), with the highest correlation being that between RPM and RPM40 (R²=1). DESeq2 and EdgeR show a correlation of R²=0.67 but no correlation between EdgeR and both RPM and RPM40.

## Raw count transformation correlation plot

```{r, fig.height=9}
df_sample1 = cbind(rpm_sample1=rpm[,1], deseq2_sample1=deseq2[,1], rpm40_sample1=rpm40[,1], edger_sample1=edger[,1])
epimedtools::et_pairs(log(df_sample1+1))
```

The transformed counts show high correlation among all normalization methods with the highest correlation being those between DESeq2, EdgeR and RPM40 (R²=1). The crushing effect appears to only apply to RPM normalization.

## Clustering consistency tables

```{r results="verbatim"}
table(gene_cluster_tab$deseq2_clust, gene_cluster_tab$edger)
table(gene_cluster_tab$rpm_clust, gene_cluster_tab$edger)
table(gene_cluster_tab$rpm40_clust, gene_cluster_tab$edger)
```

```{r, results='asis'}
rpm_deseq2_clust_table = table(gene_cluster_tab$rpm_clust, gene_cluster_tab$deseq2_clust)
rpm_edger_clust_table = table(gene_cluster_tab$rpm_clust, gene_cluster_tab$edger_clust)
rpm40_deseq2_clust_table = table(gene_cluster_tab$rpm40_clust, gene_cluster_tab$deseq2_clust)
rpm40_edger_clust_table = table(gene_cluster_tab$rpm40_clust, gene_cluster_tab$edger_clust)
deseq2_edger_clust_table = table(gene_cluster_tab$deseq2_clust, gene_cluster_tab$edger_clust)

knitr::kable(rpm_deseq2_clust_table, align = "c", caption = "RPM vs DESeq2")
knitr::kable(rpm_edger_clust_table, align = "c", caption = "RPM vs EdgeR")
knitr::kable(rpm40_deseq2_clust_table, align = "c", caption = "RPM40 vs DESeq2")
knitr::kable(rpm40_edger_clust_table, align = "c", caption = "RPM40 vs EdgeR")
knitr::kable(deseq2_edger_clust_table, align = "c", caption = "DESeq2 VS EdgeR")
```

No consistent clusters observed when comparing EdgeR to any method. Consistent clustering was only observed with DESeq2.

## Comparing patient clusters


```{r}
sample_cluster_tab = data.frame(colnames(raw_counts))
bar = cutree(rpm_hm$hc_col, k=4)
sample_cluster_tab$rpm_clust = NA
sample_cluster_tab[names(bar)[bar==1],"rpm_clust"] = "rpmA"
sample_cluster_tab[names(bar)[bar==2],"rpm_clust"] = "rpmB"
sample_cluster_tab[names(bar)[bar==3],"rpm_clust"] = "rpmC"
sample_cluster_tab[names(bar)[bar==4],"rpm_clust"] = "rpmD"
bar = cutree(deseq2_hm$hc_col, k=4)
sample_cluster_tab$deseq2_clust = NA
sample_cluster_tab[names(bar)[bar==1],"deseq2_clust"] = "deseq2A"
sample_cluster_tab[names(bar)[bar==2],"deseq2_clust"] = "deseq2B"
sample_cluster_tab[names(bar)[bar==3],"deseq2_clust"] = "deseq2C"
sample_cluster_tab[names(bar)[bar==4],"deseq2_clust"] = "deseq2D"
bar = cutree(rpm40_hm$hc_col, k=4)
sample_cluster_tab$rpm40_clust = NA
sample_cluster_tab[names(bar)[bar==1],"rpm40_clust"] = "rpm40A"
sample_cluster_tab[names(bar)[bar==2],"rpm40_clust"] = "rpm40B"
sample_cluster_tab[names(bar)[bar==3],"rpm40_clust"] = "rpm40C"
sample_cluster_tab[names(bar)[bar==4],"rpm40_clust"] = "rpm40D"
bar = cutree(edger_hm$hc_col, k=4)
sample_cluster_tab$edger_clust = NA
sample_cluster_tab[names(bar)[bar==1],"edger_clust"] = "edgerA"
sample_cluster_tab[names(bar)[bar==2],"edger_clust"] = "edgerB"
sample_cluster_tab[names(bar)[bar==3],"edger_clust"] = "edgerC"
sample_cluster_tab[names(bar)[bar==4],"edger_clust"] = "edgerD"

```

```{r results="verbatim"}
table(sample_cluster_tab$rpm_clust, sample_cluster_tab$deseq2_clust)
table(sample_cluster_tab$rpm40_clust, sample_cluster_tab$deseq2_clust)
table(sample_cluster_tab$deseq2_clust, sample_cluster_tab$edger_clust)
table(sample_cluster_tab$rpm_clust, sample_cluster_tab$edger_clust)
table(sample_cluster_tab$rpm40_clust, sample_cluster_tab$edger_clust)
```

 
# Conclusion

Clustering among the normalization methods was not consistent and genes were distributed differently among clusters. It was observed that an improvement on consistency can be done by correcting RPM values by a factor of 40, but remain inconsistent. Cluster composition was not studied but can offer an insight on the consistency. We can observe which genes switch clusters and compute the entropy between clusters. We will use a healthy tissue data set and a complex of tissue-specific genes as positive control (See `study_02_nttcga_tsps`).

It is possible to compute survival of each group cluster (?)
  
  <!-- - How to conclude? -->
  <!-- - Possible to compute **mutual information** or **entropy** between cluster. -->
  <!-- - Possible to compute survival of each groups cluster. -->

```{r, fig.width=6, fig.height=6}
load("data/lihc_surival.RData")
layout(matrix(1:4,2), respect=TRUE)
epimedtools::scurve(lihc_surival$os, paste0("grp", cutree(rpm_hm$hc_col,    k=2)[lihc_surival$id_sample]), main="rpm", xlab="time (months)")
epimedtools::scurve(lihc_surival$os, paste0("grp", cutree(deseq2_hm$hc_col, k=2)[lihc_surival$id_sample]), main="deseq2", xlab="time (months)")
epimedtools::scurve(lihc_surival$os, paste0("grp", cutree(rpm40_hm$hc_col,  k=2)[lihc_surival$id_sample]), main="rpm40", xlab="time (months)")
epimedtools::scurve(lihc_surival$os, paste0("grp", cutree(edger_hm$hc_col,  k=2)[lihc_surival$id_sample]), main="edger", xlab="time (months)")
```

## Session Info

```{r, results="verbatim"}
sessionInfo()
```
