---
title: "Impact of RNA-seq count normalization on clustering of co-expressed genes"
author: "Layla Saad Montoya and Florent Chuffart"
date: "`r Sys.Date()`"
output: 
  rmarkdown::html_document:
    toc: true
    toc_float: true
    toc_depth: 3
    number_sections: true
---

```{r, echo=FALSE, eval=TRUE, label="Loading libraries . . ."}
knitr::opts_chunk$set(collapse=TRUE, comment = "#>", fig.width=9, fig.height=6, eval=TRUE, echo=FALSE, results="hide", dpi=75)
```


# Introduction

The goal of this study is to first identify then quantify the differences between normalization methods and their subsequent gene clustering. We wish to highlight the mathematical basis for these differences as well as the ways in which these methods affect the raw results.

# Data

## Dataset

The dataset raw_counts is obtained from the NTTCGA group of TCGA database. TCGA (The Cancer Genome Atlas) is a worldwide database containing sequenced cell and tissue samples, and their respective genomic, epigenomic, transcriptomic, and proteomic data. Clincal data as well as survival data and follow-up are equally available. Our data consists of healthy, non-tumorous tissue samples, to be used as a positive control to the LIHC study (see `study_01_lihc_cpsf3`).

```{r raw_counts, echo=TRUE, results="verbatim"}
load("data/nttcga_raw_counts.RData")
raw_counts = nttcga_raw_counts
head(raw_counts[,1:4])
dim(raw_counts)
```

### Geneset

Our raw data set, containing 25360 genes, will be reduced to a gene set containing 117 tissue-specific genes belonging to lung, kidney, liver, prostate tissue. This will allow us to confirm the correlation between tissue samples and expressed genes and verify that clustering is consistent.

```{r echo=TRUE}
features = c(
  CHIAP2       = "lung",
  CHIA         = "lung",
  LOC101927884 = "lung",
  SNTN         = "lung",
  LINC02016    = "lung",
  HTR3C        = "lung",
  CSF2         = "lung",
  ROS1         = "lung",
  RPL13AP17    = "lung",
  DCSTAMP      = "lung",
  MS4A15       = "lung",
  LINC00551    = "lung",
  FAM92B       = "lung",
  LOC101928418 = "lung",
  PARAL1       = "lung",
  CT45A10      = "lung",
  CLDN19       = "kidney",
  LINC01788    = "kidney",
  LINC01762    = "kidney",
  ATP6V1G3     = "kidney",
  OR2T10       = "kidney",
  LINC01158    = "kidney",
  SLC22A13     = "kidney",
  TMEM207      = "kidney",
  LOC100506444 = "kidney",
  AFP          = "kidney",
  SLC6A18      = "kidney",
  LINC02149    = "kidney",
  CDH9         = "kidney",
  TMEM174      = "kidney",
  SLC4A9       = "kidney",
  SLC34A1      = "kidney",
  PFN3         = "kidney",
  MCCD1        = "kidney",
  SLC22A2      = "kidney",
  LINC01510    = "kidney",
  SLC7A13      = "kidney",
  LINC01230    = "kidney",
  LINC00845    = "kidney",
  HMX2         = "kidney",
  SLC22A24     = "kidney",
  SLC22A11     = "kidney",
  SLC22A12     = "kidney",
  KCNJ1        = "kidney",
  LINC00379    = "kidney",
  LINC02294    = "kidney",
  LINC00645    = "kidney",
  YWHAEP7      = "kidney",
  RDH8         = "kidney",
  SMIM34B      = "kidney",
  DNMT3L       = "kidney",
  PLAC1        = "kidney",
  PRAMEF10     = "liver",
  PRAMEF33     = "liver",
  C8A          = "liver",
  NBPF13P      = "liver",
  CFHR4        = "liver",
  CFHR2        = "liver",
  CFHR5        = "liver",
  F13B         = "liver",
  BMP10        = "liver",
  SPP2         = "liver",
  UROC1        = "liver",
  LOC100507389 = "liver",
  LINC02037    = "liver",
  LINC02428    = "liver",
  LINC02159    = "liver",
  SLC17A2      = "liver",
  SLC22A1      = "liver",
  LOC157273    = "liver",
  CYP7A1       = "liver",
  AKR1C6P      = "liver",
  GDF2         = "liver",
  MBL2         = "liver",
  F2           = "liver",
  SLC22A25     = "liver",
  SLC22A10     = "liver",
  SLC22A9      = "liver",
  LOC101929427 = "liver",
  SLCO1B1      = "liver",
  INHBE        = "liver",
  SERPINA10    = "liver",
  SERPINA2     = "liver",
  CYP1A2       = "liver",
  NPW          = "liver",
  MT1B         = "liver",
  CA5A         = "liver",
  LOC101927571 = "liver",
  C3P1         = "liver",
  CYP2A7       = "liver",
  CYP2A13      = "liver",
  APOC4        = "liver",
  FGF21        = "liver",
  HAO1         = "liver",
  LINC01370    = "liver",
  SERPINA7     = "liver",
  F9           = "liver",
  TTTY2B       = "liver",
  TTTY2        = "liver",
  CYP4F62P     = "prostate",
  CYP4F30P     = "prostate",
  LOC440910    = "prostate",
  PCGEM1       = "prostate",
  SOX14        = "prostate",
  PCAT4        = "prostate",
  SP8          = "prostate",
  RLN1         = "prostate",
  PCA3         = "prostate",
  LOC283194    = "prostate",
  LINC02297    = "prostate",
  LOC100508046 = "prostate",
  LOC101929572 = "prostate",
  OACYLP       = "prostate",
  CYP4F8       = "prostate",
  SEMG1        = "prostate",
  LINC00314    = "prostate",
  LINC00161    = "prostate"
)
idx_features = names(features)
```

```{r, fig.height=3}
plot.new()
legend("bottomleft", col=c("grey5", "grey35", "grey65", "grey95"), pch=15, paste0("tissue specific gene ", sort(unique(features))), bty="n")
legend("bottomright", col=c("grey5", "grey35", "grey65", "grey95"), pch=15, paste0("Samples ", sort(unique(substr(colnames(raw_counts), 14, 15)))), bty="n")
```

<!-- csc=c("grey5", "grey35", "grey65", "grey95")[as.numeric(as.factor(substr(colnames(raw_counts), 14, 15)))], -->
<!-- rsc=c("grey5", "grey35", "grey65", "grey95")[as.numeric(as.factor(features[rownames(data)]))], -->

# RPM

Reads Per Million (RPM) is a transcript expression unit. To calculate RPM the formula consists of dividing the number of raw counts by the total number of reads (or library size) and multiplying by a 1 million factor. This method is commonly used for inter-sample comparisons and as a basic unit of gene expression. This method does not center counts around zero. 

## Scaling factors

```{r, eval=TRUE, echo=TRUE}
rpm_sf = colSums(raw_counts)/10^6
rpm = t(t(raw_counts)/rpm_sf)
```

## Heatmap

```{r, eval=TRUE, echo=FALSE, fig.height = 9}
data = log2(rpm+1)[idx_features,]
rpm_hm = dmethr::plot_expr_hm(data, colors=c("cyan", "cyan", "cyan", "black", "red", "red", "red"), 
  normalization="zscore_rows", 
  hcmeth_rows="cor"                ,
  csc=c("grey5", "grey35", "grey65", "grey95")[as.numeric(as.factor(substr(colnames(raw_counts), 14, 15)))],
  rsc=c("grey5", "grey35", "grey65", "grey95")[as.numeric(as.factor(features[rownames(data)]))],
  main = "RPM")
```

```{r}
## Clusters
gene_cluster_tab = data.frame(features)
bar = cutree(rpm_hm$hc_row, k=4)
bar
gene_cluster_tab$rpm_clust = NA
gene_cluster_tab[names(bar)[bar==1],"rpm_clust"] = "rpmA"
gene_cluster_tab[names(bar)[bar==2],"rpm_clust"] = "rpmB"
gene_cluster_tab[names(bar)[bar==3],"rpm_clust"] = "rpmC"
gene_cluster_tab[names(bar)[bar==4],"rpm_clust"] = "rpmD"
```

# DESeq2

DESeq2 is a standard normalization method used for differential expression analysis. It is based on the RLE method (Relative Log Expression) and accounts for biases such as sequencing depth, library composition and PCR affinity / RNA composition. DESeq2 is the main normalization method used in Differential Expression Analysis at the IAB.

## Scaling factors

```{r, eval=TRUE, echo=TRUE}
colData = data.frame(id=colnames(raw_counts))
dds = DESeq2::DESeqDataSetFromMatrix(countData=raw_counts, colData=colData, design= ~ 1)
dds = DESeq2::estimateSizeFactors(dds)
deseq2_sf = dds$sizeFactor
deseq2 = t(t(raw_counts)/deseq2_sf)
deseq2 = DESeq2::counts(dds, normalized=TRUE)
```


## Heatmap

```{r, fig.height=9}
data = log2(deseq2+1)[idx_features,]
deseq2_hm = dmethr::plot_expr_hm(data, colors=c("cyan", "cyan", "cyan", "black", "red", "red", "red"), 
  normalization="zscore_rows", 
  hcmeth_rows="cor"                ,
  # hcmeth_cols = "cor",
  csc=c("grey5", "grey35", "grey65", "grey95")[as.numeric(as.factor(substr(colnames(raw_counts), 14, 15)))],
  rsc=c("grey5", "grey35", "grey65", "grey95")[as.numeric(as.factor(features[rownames(data)]))],
  main = "DESeq2"
)
```

```{r}
## Clusters
bar = cutree(deseq2_hm$hc_row, k = 4)
bar
gene_cluster_tab$deseq2_clust = NA
gene_cluster_tab[names(bar)[bar==1],"deseq2_clust"] = "deseq2A"
gene_cluster_tab[names(bar)[bar==2],"deseq2_clust"] = "deseq2B"
gene_cluster_tab[names(bar)[bar==3],"deseq2_clust"] = "deseq2C"
gene_cluster_tab[names(bar)[bar==4],"deseq2_clust"] = "deseq2D"
```

# RPM vs. DESeq2

We proceed to compare both RPM and DESeq2 methods against each other in 3 key aspects: Cluster consistency, size factor correlation and raw value transformation.

## Cluster consistency 

```{r results="verbatim"}
table(gene_cluster_tab$rpm_clust, gene_cluster_tab$deseq2_clust)
```

We observe mostly consistent gene clusters throughout the comparison (with the exception of rpmC$deseq2A, where 1 gene is displaced from each cluster), which is to be expected in the positive control. 

## Size factor correlation

```{r}
layout(matrix(1:2,1), respect=TRUE)
plot(deseq2_sf, rpm_sf)
abline(a=0, b=40, col=2)
legend("bottomright", "y=40x", col=2, lty=1)
plot(deseq2_sf, rpm_sf/40)
abline(a=0, b=1, col=2)
legend("bottomright", "y=x", col=2, lty=1)
```

In this plot, a separation into two main groups is first noticed. Two groups whose values increase at different rates but are positively correlated similar to size factor correlation in the LIHC study.

## Raw value transformation

```{r}
layout(matrix(1:2,1), respect=TRUE)
plot(log2(deseq2[,1]+1), log2(rpm[,1]+1))
plot(log2(deseq2[,1]+1), log2(40*rpm[,1]+1))
```

Finally, we can note that, when it comes to small raw values, the RPM method maintains the "crushing" effect even in the positive control. [This supports the argument that the crushing effect is inherent to RPM. (?)].

# RPM40

RPM40 counts are obtained by dividing the RPM40 size factors* by 40. This is in order to correct for standard RPM's small-value "crushing" effect.

```{r, eval=TRUE, echo=TRUE}
rpm40_sf = rpm_sf/40
rpm40 = t(t(raw_counts)/rpm40_sf)
rpm40[1:6, 1:4]
```

## Heatmap

```{r, eval=TRUE, echo=FALSE, fig.height = 9}
data = log2(rpm40+1)[idx_features,]
rpm40_hm = dmethr::plot_expr_hm(data, colors=c("cyan", "cyan", "cyan", "black", "red", "red", "red"), 
  normalization="zscore_rows", 
  hcmeth_rows="cor"                ,
  csc=c("grey5", "grey35", "grey65", "grey95")[as.numeric(as.factor(substr(colnames(raw_counts), 14, 15)))],
  rsc=c("grey5", "grey35", "grey65", "grey95")[as.numeric(as.factor(features[rownames(data)]))],
  main = "RPM40"
)
```

```{r}
## Clusters
bar = cutree(rpm40_hm$hc_row, k=4)
bar
gene_cluster_tab$rpm40_clust = NA
gene_cluster_tab[names(bar)[bar==1],"rpm40_clust"] = "rpm40A"
gene_cluster_tab[names(bar)[bar==2],"rpm40_clust"] = "rpm40B"
gene_cluster_tab[names(bar)[bar==3],"rpm40_clust"] = "rpm40C"
gene_cluster_tab[names(bar)[bar==4],"rpm40_clust"] = "rpm40D"
```

# RPM40 vs. DESeq2

## Clustering consistency table

```{r results="verbatim"}
table(gene_cluster_tab$rpm40_clust, gene_cluster_tab$deseq2_clust)
```

Clustering with RPM40 seems to have a 

# EdgeR

EdgeR is a more robust method that competes with DESeq2. EdgeR uses a TMM method to normalize counts (Trimmed Mean of M-values) and accounts for the same biases as DESeq2 such as library composition, sequencing depth and PCR affinity / RNA composition.TMM is a method that calculates size factors by first selecting a reference sample then calculating pseudo size factors using the log of the reference_sample over current sample ratio, and finally trimming the most extreme ratios and most influential genes. This way, EdgeR only takes into account differentially expressed genes and ignores the high influence of tissue-specific genes to focus normalization around housekeeping genes.

## Scaling factors

```{r, eval=TRUE, echo=TRUE}
edger_sf = edgeR::calcNormFactors(raw_counts, method="TMM")
edger = t(t(raw_counts)*edger_sf)
```

## Heatmap

```{r, eval=TRUE, echo=TRUE, fig.height = 9}
data = log2(edger+1)[idx_features,]
edger_hm = dmethr::plot_expr_hm(data, colors=c("cyan", "cyan", "cyan", "black", "red", "red", "red"), 
  normalization="zscore_rows", 
  hcmeth_rows="cor"                ,
  csc=c("grey5", "grey35", "grey65", "grey95")[as.numeric(as.factor(substr(colnames(raw_counts), 14, 15)))],
  rsc=c("grey5", "grey35", "grey65", "grey95")[as.numeric(as.factor(features[rownames(data)]))],
  main = "EdgeR")
```

```{r}
## Clusters
bar = cutree(edger_hm$hc_row, k = 4)
bar
gene_cluster_tab$edger_clust = NA
gene_cluster_tab[names(bar)[bar==1],"edger_clust"] = "edgerA"
gene_cluster_tab[names(bar)[bar==2],"edger_clust"] = "edgerB"
gene_cluster_tab[names(bar)[bar==3],"edger_clust"] = "edgerC"
gene_cluster_tab[names(bar)[bar==4],"edger_clust"] = "edgerD"
```

# Comparison
## Scaling factors ratio corr plot

```{r, fig.height=9}
df_sf = cbind(rpm_sf=rpm_sf, deseq2_sf=deseq2_sf, rpm40_sf=rpm40_sf, edger_sf=edger_sf)
epimedtools::et_pairs(df_sf)
```

In each plot excepting `rpm40_sf` vs `rpm_sf` (because corr = 1) we can observe a phenomena where the ratios of the scaling factors form two distinct groups. This is due to significant expression differences in the sample tissues. It is to note that a 40-fold factor is also present when comparing RPM to EdgeR and we can safely infer that the X-fold factor depends on the library size of the samples and is specific to the RPM computed data. 

## Raw count transformation corr plot (Sample 1)

```{r, fig.height=9}
df_sample1 = cbind(rpm_sample1=rpm[,1], deseq2_sample1=deseq2[,1], rpm40_sample1=rpm40[,1], edger_sample1=edger[,1])
epimedtools::et_pairs(log(df_sample1+1))
```

The small-value "crushing" effect is observed in all comparisons with RPM, with it being the highest when compared to EdgeR. 

## Clustering

```{r results="verbatim"}
table(gene_cluster_tab$deseq2_clust, gene_cluster_tab$edger)
table(gene_cluster_tab$rpm_clust, gene_cluster_tab$edger)
table(gene_cluster_tab$rpm40_clust, gene_cluster_tab$edger)
```

```{r, results='asis'}
rpm_deseq2_clust_table = table(gene_cluster_tab$rpm_clust, gene_cluster_tab$deseq2_clust)
rpm_edger_clust_table = table(gene_cluster_tab$rpm_clust, gene_cluster_tab$edger_clust)
edger_deseq2_clust_table = table(gene_cluster_tab$edger_clust, gene_cluster_tab$deseq2_clust)
rpm40_deseq2_clust_table = table(gene_cluster_tab$rpm40_clust, gene_cluster_tab$deseq2_clust)
rpm40_edger_clust_table = table(gene_cluster_tab$rpm40_clust, gene_cluster_tab$edger_clust)

knitr::kable(rpm_deseq2_clust_table, align = "c", caption = "RPM vs DESeq2")
knitr::kable(rpm_edger_clust_table, align = "c", caption = "RPM vs EdgeR")
knitr::kable(edger_deseq2_clust_table, align = "c", caption = "EdgeR VS DESeq2")
knitr::kable(rpm40_deseq2_clust_table, align = "c", caption = "RPM40 vs DESeq2")
knitr::kable(rpm40_deseq2_clust_table, align = "c", caption = "RPM40 vs EdgeR")
```

## Comparing patient clusters

```{r, results='asis'}
sample_cluster_tab = data.frame(colnames(raw_counts))
bar = cutree(rpm_hm$hc_col, k=4)
sample_cluster_tab$rpm_clust = NA
sample_cluster_tab[names(bar)[bar==1],"rpm_clust"] = "rpmA"
sample_cluster_tab[names(bar)[bar==2],"rpm_clust"] = "rpmB"
sample_cluster_tab[names(bar)[bar==3],"rpm_clust"] = "rpmC"
sample_cluster_tab[names(bar)[bar==4],"rpm_clust"] = "rpmD"
bar = cutree(deseq2_hm$hc_col, k=4)
sample_cluster_tab$deseq2_clust = NA
sample_cluster_tab[names(bar)[bar==1],"deseq2_clust"] = "deseq2A"
sample_cluster_tab[names(bar)[bar==2],"deseq2_clust"] = "deseq2B"
sample_cluster_tab[names(bar)[bar==3],"deseq2_clust"] = "deseq2C"
sample_cluster_tab[names(bar)[bar==4],"deseq2_clust"] = "deseq2D"
bar = cutree(rpm40_hm$hc_col, k=4)
sample_cluster_tab$rpm40_clust = NA
sample_cluster_tab[names(bar)[bar==1],"rpm40_clust"] = "rpm40A"
sample_cluster_tab[names(bar)[bar==2],"rpm40_clust"] = "rpm40B"
sample_cluster_tab[names(bar)[bar==3],"rpm40_clust"] = "rpm40C"
sample_cluster_tab[names(bar)[bar==4],"rpm40_clust"] = "rpm40D"
bar = cutree(edger_hm$hc_col, k=4)
sample_cluster_tab$edger_clust = NA
sample_cluster_tab[names(bar)[bar==1],"edger_clust"] = "edgerA"
sample_cluster_tab[names(bar)[bar==2],"edger_clust"] = "edgerB"
sample_cluster_tab[names(bar)[bar==3],"edger_clust"] = "edgerC"
sample_cluster_tab[names(bar)[bar==4],"edger_clust"] = "edgerD"

table(sample_cluster_tab$deseq2_clust, sample_cluster_tab$rpm_clust)
table(sample_cluster_tab$deseq2_clust, sample_cluster_tab$edger_clust)
table(sample_cluster_tab$rpm_clust, sample_cluster_tab$edger_clust)
table(sample_cluster_tab$rpm40_clust, sample_cluster_tab$edger_clust)
```

  - How to conclude?
  - Possible to compute **mutual information** or **enthropy** between cluster.
  - Possible to compute survival of each groups cluster.


## Session Info

```{r, results="verbatim"}
sessionInfo()
```
